src/codecoverage/bot/code_coverage_bot/zero_coverage.py

```diff
@@ -85,7 +86,71 @@ def get_fileinfo(self, filenames):
 
         return res
 
-    def generate(self, artifacts, hgrev, gitrev, out_dir='.'):
+    def generate_zero_coverage_some_platform_only(self, platforms_artifacts, hgrev, gitrev, out_dir='.'):
+        '''
+        :param platforms_artifacts: dictionary with key platform name, and value is list of artifacts
+        :param hgrev: hg revision, must be same for all reports
+        :param gitrev: git revision, must be same for all reports
+        :param out_dir: directory to output the json result
+        '''
+
+        platforms = ['linux', 'windows']
+
+        # mapping file_name -> set of platforms uncovered
+        uncovered_map = collections.defaultdict(lambda: set())
+        platform_files = collections.defaultdict(lambda: set())
+        zero_coverage_files = {}
+
+        for platform in platforms:
+            assert platform in platforms_artifacts
+
+            artifacts = platforms_artifacts[platform]
+
+            # to get all files in the artifacts
+            report = grcov.report(artifacts, out_format='coveralls+', source_dir=self.repo_dir)
+            report = json.loads(report.decode('utf-8'))
+            platform_files[platform] = {x['name'] for x in report['source_files']}
+
+            # to get zero coverage files
+            zero_coverage_report = self.generate(artifacts, hgrev, gitrev, out_dir=out_dir, writeout_result=False)
```

Comment:
This will be really slow, as we parse the artifacts once to generate the zero coverage report; then for each platform once to get platform_files, and a second time for zero_coverage_files.

In the best scenario, we should parse the artifacts only once per-platform, have per-platform reports, then pass them as input to zero_coverage_report.generate and zero_coverage_report.generate_zero_coverage_some_platform_only.

It would also be acceptable to parse the artifacts once in generate and once in generate_zero_coverage_some_platform_only, but we should avoid to parse them so many times.

Reply:
To do:
> It would also be acceptable to parse the artifacts once in generate and once in generate_zero_coverage_some_platform_only, but we should avoid to parse them so many times.

You could just refactor a part of generate out into a separate function that accepts a report as input and returns zero_coverage_report.

```
# HERE report IS THE JSON.LOADED REPORT
def get_zero_report(report):
    ...
    return zero_coverage_report

def generate(artifacts):
    # parse artifacts
    zero_coverage_report = get_zero_report(report)
    # write file
```

Then in generate_platform_only you can use get_zero_report.